packages/utility/file-server/src/caching/streamUtils.ts

```diff
@@ -0,0 +1,135 @@
+import { PassThrough, Readable } from 'stream'
+
+export interface StreamHealthMetrics {
+  isStalled: boolean
+  lastActivity: number
+  isDestroyed: boolean
+}
+
+export interface ErrorResilientStreamOptions {
+  stallTimeout?: number
+  healthCheckInterval?: number
+}
+
+interface ErrorResilientStream extends Readable {
+  healthCheck?: () => StreamHealthMetrics
+}
+
+/**
+ * Creates an error-resilient stream wrapper that prevents consumer errors from stalling the source stream
+ * Focuses on detecting and handling stalled streams
+ */
+export const createErrorResilientStream = (
+  sourceStream: Readable,
+  options: ErrorResilientStreamOptions = {},
+): ErrorResilientStream => {
+  const {
+    stallTimeout = 30000, // 30 seconds
+    healthCheckInterval = 5000, // 5 seconds
+  } = options
+
+  const passThrough: PassThrough = new PassThrough({
+    highWaterMark: 64 * 1024, // 64KB buffer to handle backpressure
+    objectMode: false,
+  })
+
+  // Simple metrics for stalled stream detection
+  const metrics: StreamHealthMetrics = {
+    isStalled: false,
+    lastActivity: Date.now(),
+    isDestroyed: false,
+  }
+
+  let healthCheckTimer: NodeJS.Timeout | null = null
+
+  // Health check function - only checks for stalled streams
+  const healthCheck = (): StreamHealthMetrics => {
+    const now = Date.now()
+    const timeSinceLastActivity = now - metrics.lastActivity
+    metrics.isStalled = timeSinceLastActivity > stallTimeout && !metrics.isDestroyed
+
+    return { ...metrics }
+  }
+
+  // Start periodic health checks for stalled streams
+  const startHealthChecks = () => {
+    if (healthCheckTimer) {
+      clearInterval(healthCheckTimer)
+    }
+
+    healthCheckTimer = setInterval(() => {
+      const health = healthCheck()
+
+      if (health.isStalled) {
+        console.warn('Stream detected as stalled:', {
+          timeSinceLastActivity: Date.now() - health.lastActivity,
+          stallTimeout,
+        })
+
+        // Destroy the stalled stream
+        passThrough.destroy(new Error('Stream stalled - no activity detected'))
+      }
+    }, healthCheckInterval)
+  }
+
+  // Handle source stream data
+  sourceStream.on('data', (chunk) => {
+    try {
+      metrics.lastActivity = Date.now()
+      metrics.isStalled = false
+
+      if (!passThrough.destroyed) {
+        passThrough.write(chunk)
+      }
```

Comment:
Should you check to see if `passThrough.write(chunk)` succeeds before updating `metrics`? Something like:

```suggestion
const writeSuccess = passThrough.write(chunk)
if (writeSuccess) {
  metrics.lastActivity = Date.now()
  metrics.isStalled = false
}
```